sdk/secp256r1/src/lib.rs

```diff
@@ -0,0 +1,615 @@
+            .map_err(|_| PrecompileError::InvalidSignature)?;
+
+        // Parse order - 1 from constant
+        let order_minus_one: BigNum = BigNum::from_slice(&SECP256R1_ORDER_MINUS_ONE)
+            .map_err(|_| PrecompileError::InvalidSignature)?;
+
+        // Create a BigNum for 1
+        let one = BigNum::from_u32(1).map_err(|_| PrecompileError::InvalidSignature)?;
+
+        // Define curve group
+        let group = EcGroup::from_curve_name(Nid::X9_62_PRIME256V1)
+            .map_err(|_| PrecompileError::InvalidSignature)?;
+        let mut ctx = BigNumContext::new().map_err(|_| PrecompileError::InvalidSignature)?;
+
+        for i in 0..num_signatures {
+            let start = i
+                .saturating_mul(SIGNATURE_OFFSETS_SERIALIZED_SIZE)
+                .saturating_add(SIGNATURE_OFFSETS_START);
+            let end = start.saturating_add(SIGNATURE_OFFSETS_SERIALIZED_SIZE);
+
+            // bytemuck wants structures aligned
+            let offsets: &Secp256r1SignatureOffsets =
+                bytemuck::try_from_bytes(&data[start..end])
+                    .map_err(|_| PrecompileError::InvalidDataOffsets)?;
+
+            // Parse out signature
+            let signature = get_data_slice(
+                data,
+                instruction_datas,
+                offsets.signature_instruction_index,
+                offsets.signature_offset,
+                SIGNATURE_SERIALIZED_SIZE,
+            )?;
+
+            // Parse out pubkey
+            let pubkey = get_data_slice(
+                data,
+                instruction_datas,
+                offsets.public_key_instruction_index,
+                offsets.public_key_offset,
+                COMPRESSED_PUBKEY_SERIALIZED_SIZE,
+            )?;
+
+            // Parse out message
+            let message = get_data_slice(
+                data,
+                instruction_datas,
+                offsets.message_instruction_index,
+                offsets.message_data_offset,
+                offsets.message_data_size as usize,
+            )?;
+
+            let r_bignum = BigNum::from_slice(&signature[..32])
+                .map_err(|_| PrecompileError::InvalidSignature)?;
+            let s_bignum = BigNum::from_slice(&signature[32..])
+                .map_err(|_| PrecompileError::InvalidSignature)?;
+
+            // Check that r and s are within the proper range and enforce a low-S value
+            let within_range = r_bignum > one
+                && r_bignum <= order_minus_one
+                && s_bignum > one
+                && s_bignum <= order_minus_one
+                && s_bignum <= half_order;
+
+            if !within_range {
+                return Err(PrecompileError::InvalidSignature);
+            }
+
+            // Create an ECDSA signature object from the ASN.1 integers
+            let ecdsa_sig = openssl::ecdsa::EcdsaSig::from_private_components(r_bignum, s_bignum)
+                .map_err(|_| PrecompileError::InvalidSignature)?;
```

Comment:
```suggestion
            let ecdsa_sig = openssl::ecdsa::EcdsaSig::from_private_components(r_bignum, s_bignum)
                .and_then(|s| s.to_der())
                .map_err(|_| PrecompileError::InvalidSignature)?;
```
This is very nit, but the verification logic in lines 269 - 278 is a bit messy to read, so maybe moving `to_der()` up here will simplify it a little.